Add MQTT climate setting for current humidity

Stonos opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This change introduces two new variables to the MQTT Climate: `current_humidity_template` and `current_humidity_topic`. These allow you to set the current humidity of the MQTT Climate, and work in a similar way to `current_temperature_template` and `current_temperature_topic` respectively.

One benefit of this is that it allows other integrations to use this value. For example, it's possible to see the current humidity in Google Home with this change:

![image](https://user-images.githubusercontent.com/42659/209599002-41e5323c-1bc0-4cb3-9310-43a835c11a67.png)


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/25413

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Stonos committed:
MQTT Climate: Add support for setting the current humidity via MQTT

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -216,6 +220,8 @@ def valid_preset_mode_configuration(config: ConfigType) -> ConfigType:
         vol.Optional(CONF_AUX_STATE_TEMPLATE): cv.template,
         vol.Optional(CONF_AUX_STATE_TOPIC): valid_subscribe_topic,
+        vol.Optional(CONF_CURRENT_HUMIDITY_TEMPLATE): cv.template,
+        vol.Optional(CONF_CURRENT_HUMIDITY_TOPIC): valid_subscribe_topic,
```
Do we not need a `CONF_HUMIDITY_COMMAND_TOPIC` too? Further it would be good to add `CONF_MIN_HUMIDITY` and `CONF_MAX_HUMIDITY` as well.

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -601,6 +607,18 @@ def handle_temperature_high_received(msg: ReceiveMessage) -> None:
+        @log_messages(self.hass, self.entity_id)
+        def handle_current_humidity_received(msg: ReceiveMessage) -> None:
+            """Handle current humidity coming via MQTT."""
+            handle_temperature_received(
```
The name of the function still says it handles a temperature received.

jbouwh reviewed (changes requested):
Nice idea to add this missing functionality to MQTT climate. It only misses support to be able to set a humidity yet.

jbouwh commented:
Let me know if you can follow up on the request to do some changes. I think the start is okay. If you like , I could help getting some progress to this Idea. Let me know, what you if you would like that.

Stonos commented:
Sorry, I didn't have much time to work on it in the past few days. I'll try to make the requested changes either today or tomorrow. Thanks, I'll let you know if I have any questions/need help!

Stonos committed:
MQTT Climate: Add configuration constants related to setting the target humidity

Stonos committed:
MQTT Climate: Add support for setting the humidity's state topic & template

Stonos committed:
MQTT Climate: Add support for setting the initial humidity

Stonos committed:
MQTT Climate: Add support for setting the humidity's command topic & template

Stonos committed:
MQTT Climate: Add support for setting the min/max humidity

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -226,6 +245,13 @@ def valid_preset_mode_configuration(config: ConfigType) -> ConfigType:
         vol.Optional(CONF_FAN_MODE_STATE_TOPIC): valid_subscribe_topic,
+        vol.Optional(CONF_HUMIDITY_COMMAND_TEMPLATE): cv.template,
+        vol.Optional(CONF_HUMIDITY_COMMAND_TOPIC): valid_publish_topic,
+        vol.Optional(CONF_HUMIDITY_INITIAL, default=50): cv.positive_int,
```
Why would we need an Initial value. I would suggest assigning `None` if a state topic or command topic is set during initialisation.

jbouwh replied:
I think it is not needed to have an initial value. Initial values can be reported by the device. It might even be that I will sugest to remove those initial settings and deprecate them.
Even the `None` assignments for all attributes (https://github.com/home-assistant/core/pull/84592/files#diff-c8ebf96494fa4dca0f6c463949933572a837d982d5d9fbd8f042173bba03fbd1R466-R473) can be cleaned up since they are already initialized in the parent class. We do not need to reset those to `None` when a config update arrives.
You do not need to changed that in this PR though, I'll open a seperate one for that.

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -226,6 +245,13 @@ def valid_preset_mode_configuration(config: ConfigType) -> ConfigType:
+        vol.Optional(CONF_HUMIDITY_COMMAND_TEMPLATE): cv.template,
+        vol.Optional(CONF_HUMIDITY_COMMAND_TOPIC): valid_publish_topic,
+        vol.Optional(CONF_HUMIDITY_INITIAL, default=50): cv.positive_int,
+        vol.Optional(CONF_HUMIDITY_MIN, default=DEFAULT_MIN_HUMIDITY): vol.Coerce(float),
```
Can we add validation in min and max humidity like MQTT humidifier?

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -753,23 +813,23 @@ async def async_set_temperature(self, **kwargs: Any) -> None:
         )
 
-        await self._set_temperature(
+        await self._set_mqtt_attribute(
```
May be `_set_climate_attribute` is better?

Stonos committed:
MQTT Climate: Fix style & tests

Stonos committed:
MQTT Climate: Set the initial humidity to None

Stonos committed:
MQTT Climate: Rename _set_mqtt_attribute to _set_climate_attribute and handle_temperature_received to handle_climate_attribute_received

Stonos committed:
MQTT Climate: Copy humidity range validation from MQTT Humidifier

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -92,6 +96,13 @@
+CONF_HUMIDITY_COMMAND_TOPIC = "target_humidity_command_topic"
+CONF_HUMIDITY_STATE_TEMPLATE = "target_humidity_state_template"
+CONF_HUMIDITY_STATE_TOPIC = "target_humidity_state_topic"
+CONF_HUMIDITY_INITIAL = "initial_humidity"
```
```suggestion
```

I think this one can be cleaned up

jbouwh commented on homeassistant/components/mqtt/abbreviations.py:
```diff
@@ -83,6 +85,7 @@
     "ic": "icon",
     "img_e": "image_encoding",
     "init": "initial",
+    "init_hum": "initial_humidity",
```
```suggestion
    "init_hum": "initial_humidity",
```
Can be cleaned up

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -226,6 +255,17 @@ def valid_preset_mode_configuration(config: ConfigType) -> ConfigType:
         vol.Optional(CONF_FAN_MODE_STATE_TOPIC): valid_subscribe_topic,
+        vol.Optional(CONF_HUMIDITY_COMMAND_TEMPLATE): cv.template,
+        vol.Optional(CONF_HUMIDITY_COMMAND_TOPIC): valid_publish_topic,
+        vol.Optional(CONF_HUMIDITY_INITIAL, default=None): cv.positive_int,
```
```suggestion
```

Do not think we need this

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -412,6 +456,9 @@ def _setup_from_config(self, config: ConfigType) -> None:
         if self._topic[CONF_TEMP_HIGH_STATE_TOPIC] is None:
             self._attr_target_temperature_high = config[CONF_TEMP_INITIAL]
 
+        if self._topic[CONF_HUMIDITY_STATE_TOPIC] is None:
+            self._attr_target_humidity = config[CONF_HUMIDITY_INITIAL]
+
```
```suggestion
```
At `self._attr_target_humidity = None` at self.__init__() instead if `CONF_HUMIDITY_STATE_TOPIC` is `None`

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -472,6 +519,11 @@ def _setup_from_config(self, config: ConfigType) -> None:
         ):
             support |= ClimateEntityFeature.TARGET_TEMPERATURE_RANGE
 
+        if (self._topic[CONF_HUMIDITY_STATE_TOPIC] is not None) or (
+            self._topic[CONF_HUMIDITY_COMMAND_TOPIC] is not None
+        ):
```
```suggestion
        if self._topic[CONF_HUMIDITY_COMMAND_TOPIC] is not None:
```
We can also set the target humidity if  there is no state topic. In that case it will be set in optimistic mode.

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -593,14 +645,38 @@ def handle_temperature_low_received(msg: ReceiveMessage) -> None:
+        @log_messages(self.hass, self.entity_id)
+        def handle_current_humidity_received(msg: ReceiveMessage) -> None:
+            """Handle current humidity coming via MQTT."""
+            handle_climate_attribute_received(
```
Just for the record
It is okay to enable to setting the current humidity attribute, even if `ClimateEntityFeature.TARGET_HUMIDITY` is not set.

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -593,14 +645,38 @@ def handle_temperature_low_received(msg: ReceiveMessage) -> None:
+        @log_messages(self.hass, self.entity_id)
+        def handle_target_humidity_received(msg: ReceiveMessage) -> None:
+            """Handle target humidity coming via MQTT."""
+            handle_climate_attribute_received(
```
I think we should check here if `ClimateEntityFeature.TARGET_HUMIDITY` is set. It is some weird that we can receive the target humidity but not set it.

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -226,6 +255,17 @@ def valid_preset_mode_configuration(config: ConfigType) -> ConfigType:
+            float
+        ),
+        vol.Optional(CONF_HUMIDITY_STATE_TEMPLATE): cv.template,
+        vol.Optional(CONF_HUMIDITY_STATE_TOPIC): valid_subscribe_topic,
```
I suggest we add a check that `CONF_HUMIDITY_STATE_TOPIC` can only be configured if also `CONF_HUMIDITY_COMMAND_TOPIC` is configured.

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -779,6 +855,19 @@ async def async_set_temperature(self, **kwargs: Any) -> None:
+    async def async_set_humidity(self, humidity: int) -> None:
+        """Set new target humidity."""
+
+        await self._set_climate_attribute(
```
We should check for `ClimateEntityFeature.TARGET_HUMIDITY` to be set here and raise some error if this is not supported.

jbouwh reviewed (commented):
Nice progress, added some more comments.

Stonos committed:
MQTT Climate: Remove CONF_HUMIDITY_INITIAL

Stonos committed:
MQTT Climate: Only enable support for TARGET_HUMIDITY when the command topic is set

Stonos committed:
MQTT Climate: Check if setting the target humidity is supported before actually setting it

Stonos committed:
MQTT Climate: Make sure that CONF_HUMIDITY_COMMAND_TOPIC has been configured when setting CONF_HUMIDITY_STATE_TOPIC

jbouwh commented:
I had https://github.com/home-assistant/core/pull/84777 merged. This PR adds optimistic mode support to MQTT climate. You will have to rebase your PR.

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -541,7 +593,7 @@ def handle_action_received(msg: ReceiveMessage) -> None:
 
         @callback
-        def handle_temperature_received(
+        def handle_climate_attribute_received(
```
The only line that seems to be de duplicated seems the log line, which is stale for `humidity`. May be we should just assign the attributes value in the specific handlers and dispose this one.

jbouwh replied:
Line 627 still shows `Could not parse temperature from`.
This is not applicable for humidity attributes. As I already commented it might be better not using an extra function for setting attributes. Directly setting them perhaps would be more directly and clear.

Stonos replied:
Looks like I had forgotten to submit my pending comments, so you weren't able to see my comment. Can you take a look at my suggestion above?

jbouwh replied:
> Could not parse temperature_state_template from [...]

I think that is a good alternative!

Stonos committed:
MQTT Climate: Fix broken tests

Stonos committed:
MQTT Climate: Add test for optimistically setting the target humidity

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -99,6 +103,12 @@
 CONF_HOLD_STATE_TOPIC = "hold_state_topic"
 CONF_HOLD_LIST = "hold_modes"
 
+CONF_HUMIDITY_COMMAND_TEMPLATE = "target_humidity_command_template"
+CONF_HUMIDITY_COMMAND_TOPIC = "target_humidity_command_topic"
+CONF_HUMIDITY_STATE_TEMPLATE = "target_humidity_state_template"
+CONF_HUMIDITY_STATE_TOPIC = "target_humidity_state_topic"
```
May be we can leave the `target_` prefix away. The topics get much the same like the temperature topics. I know I asked you at first to add the prefix.

Stonos replied:
That's what I tried to do at the beginning, but when I went to add them to `abbreviations.py`, I found out that they already exist with the `target_` prefix (by the MQTT Humidifier): https://github.com/home-assistant/core/blob/c56806a676e8c2482c6ee9e5880c2d2f01af59ef/homeassistant/components/mqtt/abbreviations.py#L88-L91

We could drop the `target_` prefix, but then we'd have to come up with a new abbreviation.

jbouwh replied:
Then lets leave it the way it is!

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -606,14 +666,46 @@ def handle_temperature_low_received(msg: ReceiveMessage) -> None:
+        def handle_target_humidity_received(msg: ReceiveMessage) -> None:
+            """Handle target humidity coming via MQTT."""
+
+            if not self.supported_features & ClimateEntityFeature.TARGET_HUMIDITY:
```
You need some additional tests to ensure code test coverage here.

Stonos replied:
I added a test for `handle_target_humidity_received`, but it only tests the happy path. I don't think that it's possible for `handle_target_humidity_received` to be called without having the `TARGET_HUMIDITY` feature ever since I added this validation check: 

https://github.com/Stonos/core/blob/81ca8dcb1779ac2a8f92c3f9e51831e849492eb5/homeassistant/components/mqtt/climate.py#L247-L257

Perhaps we should remove the check from `handle_target_humidity_received`?

jbouwh replied:
Sound good to me!

jbouwh reviewed (commented):
Almost there, nice work! Left a few comments.

Stonos committed:
MQTT Climate: Remove references to "temperature" in handle_climate_attribute_received

Stonos committed:
MQTT Climate: Add additional humidity-related tests

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -606,14 +666,46 @@ def handle_temperature_low_received(msg: ReceiveMessage) -> None:
+        def handle_target_humidity_received(msg: ReceiveMessage) -> None:
+            """Handle target humidity coming via MQTT."""
+
+            if not self.supported_features & ClimateEntityFeature.TARGET_HUMIDITY:
+                _LOGGER.error(
+                    "Attempted to set target humidity with no %s configured",
+                    CONF_HUMIDITY_COMMAND_TOPIC,
+                )
+                return
+
```
```suggestion
```

jbouwh reviewed (commented):
Looks nice, I left a suggestion for you to remove the validation on receiving the state target humidity as it is not needed.
I'll ask @emontnemery to have a look as well.

Stonos committed:
MQTT Climate: Remove supported feature check in handle_target_humidity_received

It's not needed because this is covered by the `valid_humidity_state_configuration` validation.

Co-authored-by: Jan Bouwhuis <jbouwh@users.noreply.github.com>

emontnemery commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -798,6 +883,26 @@ async def async_set_temperature(self, **kwargs: Any) -> None:
+    async def async_set_humidity(self, humidity: int) -> None:
+        """Set new target humidity."""
+
+        if not self.supported_features & ClimateEntityFeature.TARGET_HUMIDITY:
+            _LOGGER.error(
+                "Attempted to set target humidity with no %s configured",
+                CONF_HUMIDITY_COMMAND_TOPIC,
+            )
+            return
```
I don't think this is needed, the base entity already checks if setting humidity is supported: https://github.com/home-assistant/core/blob/db3e57542558cf06c2ec9c7f009626466569b0d2/homeassistant/components/climate/__init__.py#L170-L176

jbouwh replied:
Added a sugestion to remove the check.

jbouwh commented on homeassistant/components/mqtt/climate.py:
```diff
@@ -798,6 +883,26 @@ async def async_set_temperature(self, **kwargs: Any) -> None:
 
+    async def async_set_humidity(self, humidity: int) -> None:
+        """Set new target humidity."""
+
+        if not self.supported_features & ClimateEntityFeature.TARGET_HUMIDITY:
+            _LOGGER.error(
+                "Attempted to set target humidity with no %s configured",
+                CONF_HUMIDITY_COMMAND_TOPIC,
+            )
+            return
```
```suggestion
```

Stonos committed:
MQTT Climate: Remove supported feature check in async_set_humidity

It is covered by the base Climate entity.

Co-authored-by: Jan Bouwhuis <jbouwh@users.noreply.github.com>

jbouwh reviewed (approved):
Nice first contribution,
Thnx @Stonos!

Stonos commented:
Thank you! And thanks for the review & the guidance you've provided! 🙂

I will update the documentation next.
